cmd/oras/root/manifest/index/create.go

```diff
@@ -0,0 +1,207 @@
+	ctx, logger := command.GetLogger(cmd, &opts.Common)
+	target, err := opts.NewTarget(opts.Common, logger)
+	if err != nil {
+		return err
+	}
+	// if a digest is given as the index reference, we need to ignore it
+	opts.RawReference, _, _ = strings.Cut(opts.RawReference, "@")
+	// we assume that the sources and the to be created index are all in the same
+	// repository, so no copy is needed
+	manifests, err := resolveSourceManifests(ctx, target, opts)
+	if err != nil {
+		return err
+	}
+	desc, content, err := packIndex(&ocispec.Index{}, manifests)
+	if err != nil {
+		return err
+	}
+	opts.Println("Created the index")
+	return pushIndex(ctx, target, desc, content, opts.Reference, opts.extraRefs, opts.AnnotatedReference(), opts.Printer)
+}
+
+func resolveSourceManifests(ctx context.Context, target oras.ReadOnlyTarget, opts createOptions) ([]ocispec.Descriptor, error) {
+	var resolved []ocispec.Descriptor
+	for _, source := range opts.sources {
+		desc, content, err := oras.FetchBytes(ctx, target, source, oras.DefaultFetchBytesOptions)
+		if err != nil {
+			if errors.Is(err, errdef.ErrNotFound) {
+				return nil, fmt.Errorf("could not resolve %s. Make sure the manifest exists in %s", source, opts.Path)
+			}
+			return nil, err
+		}
+		opts.Println("Resolved manifest", source)
+		if descriptor.IsImageManifest(desc) {
+			desc.Platform, err = getPlatform(ctx, target, content)
+			if err != nil {
+				return nil, err
+			}
+		}
+		resolved = append(resolved, desc)
+	}
+	return resolved, nil
+}
+
+func getPlatform(ctx context.Context, target oras.ReadOnlyTarget, manifestBytes []byte) (*ocispec.Platform, error) {
+	// extract config descriptor
+	var manifest ocispec.Manifest
+	if err := json.Unmarshal(manifestBytes, &manifest); err != nil {
+		return nil, err
+	}
+	// fetch config content
+	contentBytes, err := content.FetchAll(ctx, target, manifest.Config)
+	if err != nil {
+		return nil, err
+	}
+	var platform ocispec.Platform
+	if err := json.Unmarshal(contentBytes, &platform); err != nil {
+		return nil, err
+	}
+	return &platform, nil
+}
+
+func packIndex(oldIndex *ocispec.Index, manifests []ocispec.Descriptor) (ocispec.Descriptor, []byte, error) {
+	index := ocispec.Index{
+		Versioned: specs.Versioned{
+			SchemaVersion: 2,
+		},
+		MediaType:    ocispec.MediaTypeImageIndex,
+		ArtifactType: oldIndex.ArtifactType,
+		Manifests:    manifests,
+		Subject:      oldIndex.Subject,
+		Annotations:  oldIndex.Annotations,
+	}
+	indexBytes, err := json.Marshal(index)
+	if err != nil {
+		return ocispec.Descriptor{}, nil, err
+	}
+	desc := content.NewDescriptorFromBytes(ocispec.MediaTypeImageIndex, indexBytes)
+	return desc, indexBytes, nil
+}
+
+func pushIndex(ctx context.Context, target oras.Target, desc ocispec.Descriptor, content []byte, ref string, extraRefs []string, path string, printer *output.Printer) error {
+	// push the index
+	var err error
+	if ref == "" || contentutil.IsDigest(ref) {
```

Comment:
`contentutil.IsDigest(ref)` is a bad path and should be avoided (Image a user can predict the digest of packed index, they don't need `oras` at all)